Adds GCP-based Jenkins test

diverdane opened the pull request:
### What does this PR do?
This change adds GCP/GKE based Jenkins tests that do the following
using Helm v3:
 - Creates a Kubernetes namespace
 - Runs a Helm install of a Conjur OSS server
 - Runs a Helm test that deploys a test container that
   confirms that the Conjur server's status page is active.
 - Does a Helm delete of the Conjur OSS release
 - Deletes the namespace

This change Also fixes an issue with running the 'test.sh' script with Helm v2
in that the 'helm test ...' and 'helm delete ...' commands were
incorrectly including a namespace argument (a namespace argument
is only used for 'helm install ...' for Helm v2).

This Jenkins test requires a change in the jenkins-seed repo:
    https://github.com/conjurinc/jenkins-seed/pull/72

### What ticket does this PR close?
Issue #65 

### Checklists

#### Change log
- [ ] The CHANGELOG has been updated, or
- [x] This PR does not include user-facing changes and doesn't require a CHANGELOG update

#### Test coverage
- [x] This PR includes new unit and integration tests to go with the code changes, or
- [ ] The changes in this PR do not require tests

#### Documentation
- [ ] Docs (e.g. `README`s) were updated in this PR, and/or there is a follow-on issue to update docs, or
- [x] This PR does not require updating any documentation

sgnn7 commented on Jenkinsfile:
```diff
@@ -0,0 +1,48 @@
+      stage('GKE Build and Test with Helm v2') {
+        steps {
+          sh 'cd ci && summon ./jenkins_build.sh'
+          script {
```
Shouldn't this be `environment`? Same for the block below.

sgnn7 commented on ci/Dockerfile:
```diff
@@ -0,0 +1,32 @@
+FROM google/cloud-sdk
+ARG HELM_VERSION=3.1.3
```
nit: Add a newline above this one for readability

sgnn7 commented on ci/Dockerfile:
```diff
@@ -0,0 +1,32 @@
+                       gnupg2 \
+                       software-properties-common \
+                       wget && \
+    curl -fsSL https://download.docker.com/linux/$(. /etc/os-release; echo "$ID")/gpg | apt-key add - && \
```
Download/save the key and then import rather than piping - it doesn't generate warnings and is less fragile

diverdane replied:
Fixed.

sgnn7 commented on utils.sh:
```diff
@@ -0,0 +1,20 @@
+    echo "$1"
+    echo "==================================================="
+}
+
```
Stray newline

sgnn7 commented on ci/platform_login.sh:
```diff
@@ -0,0 +1,14 @@
+gcloud container clusters get-credentials \
+  $GCLOUD_CLUSTER_NAME \
+  --zone $GCLOUD_ZONE \
+  --project $GCLOUD_PROJECT_NAME
```
All of the vars should be quoted in the file

sgnn7 commented on ci/jenkins_build.sh:
```diff
@@ -0,0 +1,126 @@
+export KUBECTL_VERSION="${KUBECTL_VERSION:-1.16.9}"
+export SKIP_GCLOUD_LOGIN="${SKIP_GCLOUD_LOGIN:-false}"
+# Use release name not based on timestamp since tests can run in parallel
+export RELEASE_NAME="$CONJUR_NAMESPACE"
```
Sort these alphabetically - they're really hard to read as-is

sgnn7 commented on ci/jenkins_build.sh:
```diff
@@ -0,0 +1,126 @@
+#                         access to a cluster. Defaults to 'false'.
+
+test_id="$(random_string)"
+export CONJUR_NAMESPACE="${CONJUR_NAMESPACE:-conjur-oss-test-$test_id}"
```
Add newline above this for readability

sgnn7 commented on ci/jenkins_build.sh:
```diff
@@ -0,0 +1,126 @@
+
+announce "Building gcloud/kubectl/helm client image..."
+tmp_dir="$(pwd)/.tmp"
+bin_dir="${tmp_dir}/bin"
```
For maintainability, this should indicate more what it is (maybe `tmp_bin_dir`?)

sgnn7 commented on ci/jenkins_build.sh:
```diff
@@ -0,0 +1,126 @@
+bin_dir="${tmp_dir}/bin"
+tools_image_name="conjur-oss-helm-kubectl"
+
+mkdir -p "${bin_dir}"
+mkdir -p "${tmp_dir}/.kube"
+mkdir -p "${tmp_dir}/.config"
```
```suggestion
mkdir -p "${bin_dir}" \
         "${tmp_dir}/.kube" \
         "${tmp_dir}/.config"
```

sgnn7 commented on ci/jenkins_build.sh:
```diff
@@ -0,0 +1,126 @@
+             --build-arg HELM_VERSION="$HELM_VERSION" \
+             --build-arg KUBECTL_VERSION="$KUBECTL_VERSION" \
+             -f Dockerfile \
+             .
```
This image should be built before you make the executable /bin links probably

sgnn7 commented on ci/jenkins_build.sh:
```diff
@@ -0,0 +1,126 @@
+  announce "                 SUCCESS"
+else
+  announce "                 FAILED"
+fi
```
This if/else block will never run - you have the `set -e` flag which will make `./test.sh` exit the script on failure. Even if it did run, the error would not propagate to jenkins.

diverdane commented:
Hi @sgnn7 I've addressed your comments, it's ready for re-review.
One big change I made is to delete the test for Helm V2. To get this to work, it would probably just take a `helm init --upgrade` on our GKE testbed to update Helm V2 to v2.16 (the current version is v2.10.0, and this version is incompatible with the Kubernetes V1.16 that is running on this cluster. However, including a test for Helm V2 would mean we'd have to maintain Tiller on this cluster. We already have a Helm V2 test based on KinD, so testing on GKE is not absolutely necessary.
